### PR TITLE
feat: GQL tax rates

### LINKED ISSUE #5761: [GQL] Tax Rates (included/taxes-rates)
### Collections

- [ ] Taxes

### Meteor dependencies

- [ ] meteor/check
- [ ] meteor/meteor

### Methods

- [ ] taxes/addRate
- [ ] taxes/editRate
- [ ] taxes/deleteRate

### Autoform fields

- [ ] shopId
- [ ] country
- [ ] region
- [ ] postal
- [ ] taxCode
- [ ] rate

### Publications

- [ ] Taxes (finds the list of taxes by shopId)

### File structure

```
.
├── index.js
├── methods.js
└── publications.js
```

### PR DESCRIPTION
**Pull request by mikemurray**
Resolves #5761  
Impact: **major**  
Type: **feature**

## Issue

No GQL queries and mutation for tax rates.

## Solution

Add GQL queries and mutations for tax rates.

## Testing
1. Add a tax rate
1. Update a tax rate
1. Query for tax rates
1. Delete a tax rate
1. Query again for tax rates and ensure the one you added is gone

### Here are the mutations/queries, you'll need to update the IDs to your shop and tax rate documents.

```graphql
query Queries {
  primaryShopId
  taxRates(shopId: "cmVhY3Rpb24vc2hvcDpDN2NNS3ptOXVUVGtHbnNxcw==") {
		nodes {
      _id
    	rate
      shop {
        _id
      }
      postal
      country
      taxCode
      postal
    }
  }
}

mutation AddTaxRate {
  createTaxRate(input: {
    region: "CA",
    rate: 0.10,
    shopId: "cmVhY3Rpb24vc2hvcDpDN2NNS3ptOXVUVGtHbnNxcw==",
    country: "USA",
    postal: "90066",
    taxCode: "what?"
  }) {
    taxRate {
      _id
      rate
      shop {
        _id
      }
      postal
      country
      taxCode
      postal
    }
  }
}

mutation UpdateTaxRate {
  updateTaxRate(input: {
    taxRateId: "cmVhY3Rpb24vdGF4UmF0ZTpvcEY4VGg5ZDJUS0Rqcm5yOQ==",
    shopId: "cmVhY3Rpb24vc2hvcDpDN2NNS3ptOXVUVGtHbnNxcw==",
    region: "CA",
    rate: 0.40,
    country: "USA",
    postal: "90066",
    taxCode: "what?"
  }) {
    taxRate {
      _id
      rate
      shop {
        _id
      }
      postal
      country
      taxCode
      postal
    }
  }
}

mutation DeleteTaxRate {
  deleteTaxRate(input: {
    taxRateId: "cmVhY3Rpb24vdGF4UmF0ZTo0YWc1OUdNOXl0WGNnTGlwRw==",
    shopId: "cmVhY3Rpb24vc2hvcDpDN2NNS3ptOXVUVGtHbnNxcw=="
  }) {
    taxRate {
      _id
      rate
      shop {
        _id
      }
      postal
      country
      taxCode
      postal
    }
  }
}
```

### CODE REVIEW

**Comment by mikemurray** [comment]
@willopez @aldeed ready for review

File: src/plugins/taxes-rates/mutations/createTaxRate.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    _id: Random.id(),
+    shopId,
+    country,
+    region,
+    postal,
+    taxCode,
+    rate
+  };
+
+  await TaxRates.insert(taxRate);
```
Comment (willopez) [inline src/plugins/taxes-rates/mutations/createTaxRate.js]: method deprecated, use `insertOne` https://mongodb.github.io/node-mongodb-native/3.3/api/Collection.html#insert

File: src/plugins/taxes-rates/mutations/deleteTaxRate.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  const { TaxRates } = collections;
+
+  await checkPermissions(["admin", "owner"], shopId);
+
+  const taxRateToDelete = await TaxRates.findOne({
+    _id,
+    shopId
+  });
+
+  await TaxRates.remove({
```
Comment (willopez) [inline src/plugins/taxes-rates/mutations/deleteTaxRate.js]: method deprecated use `removeOne` https://mongodb.github.io/node-mongodb-native/3.3/api/Collection.html#remove

File: src/plugins/taxes-rates/resolvers/Mutation/createTaxRate.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { decodeShopOpaqueId } from "../../xforms/id.js";
```
Comment (willopez) [inline src/plugins/taxes-rates/resolvers/Mutation/createTaxRate.js]: @aldeed what is the recommended pattern to decode opaque ids? importing utility functions in this way, seems to couple plugins with core. How about importing from `import decodeOpaqueIdForNamespace from "@reactioncommerce/api-utils/decodeOpaqueIdForNamespace.js";` and using ` const shopId = decodeOpaqueIdForNamespace("reaction/shop", opaqueShopId);` ?

File: src/plugins/taxes-rates/resolvers/Mutation/deleteTaxRate.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { decodeShopOpaqueId, decodeTaxRateOpaqueId } from "../../xforms/id.js";
```
Comment (willopez) [inline src/plugins/taxes-rates/resolvers/Mutation/deleteTaxRate.js]: same questions as above

File: src/plugins/taxes-rates/schemas/schema.graphql [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+"Describes the input for creating a tax rate"
+input CreateTaxRateInput {
+  "An optional string identifying the mutation call, which will be returned in the response payload"
+  clientMutationId: String
+
+  "Did this tax match due to the order origin or the order destination?"
+  country: String!
+
+  "A human-readable display name for showing this tax to operators and customers in the UI"
```
Comment (willopez) [inline src/plugins/taxes-rates/schemas/schema.graphql]: is this description correct?

File: src/plugins/taxes-rates/schemas/schema.graphql [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    input: DeleteTaxRateInput!
+  ): DeleteTaxRatePayload
+}
+
+"A connection edge in which each node is a `TaxRate` object"
+type TaxRateEdge {
+  "The cursor that represents this node in the paginated results"
+  cursor: ConnectionCursor!
+
+  "The catalog item"
```
Comment (willopez) [inline src/plugins/taxes-rates/schemas/schema.graphql]: did you mean, the tax rate? this is not for catalog items

File: tests/integration/api/mutations/taxRates/taxRates.test.js [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    region: "CA",
+    rate: 0.10,
+    country: "USA",
+    postal: "90066",
+    taxCode: "CODE"
+  };
+
+  expect(createdTaxRate).toEqual(expectedTaxRateResponse);
+
+  // Check the database for the new TaxRate document
```
Comment (willopez) [inline tests/integration/api/mutations/taxRates/taxRates.test.js]: is this double-check really necessary? the GraphQL operation returns the created/updated document, seems redundant to query the db, unless I am missing something?

File: src/plugins/taxes-rates/resolvers/Mutation/createTaxRate.js (hunk 3 above)
Comment (mikemurray) [reply src/plugins/taxes-rates/resolvers/Mutation/createTaxRate.js]: @willopez This imports from its own package. The id.js file imports from "api-utils". This is the same pattern used in the other plugins.

File: src/plugins/taxes-rates/resolvers/Mutation/deleteTaxRate.js (hunk 4 above)
Comment (mikemurray) [reply src/plugins/taxes-rates/resolvers/Mutation/deleteTaxRate.js]: @willopez same answer as the one I posted above

File: tests/integration/api/mutations/taxRates/taxRates.test.js (hunk 7 above)
Comment (mikemurray) [reply tests/integration/api/mutations/taxRates/taxRates.test.js]: The GQL mutation/query returns a resolved document, that may not be what's actually in the database. For example, the field `shop` is available in queries, but only `shopId` is stored on the document in the DB. A resolver may add additional fields or formatting to a field.

The query to the database is to ensure the document was stored how it's expected to be before any GQL resolvers manipulate it.

**Comment by mikemurray** [comment]
@willopez Addressed comments and ready for a re-review.

File: tests/integration/api/mutations/taxRates/taxRates.test.js (hunk 7 above)
Comment (willopez) [reply tests/integration/api/mutations/taxRates/taxRates.test.js]: make sense

**Pull request merged**